src/Shared/Components/ComponentMarker.cs

```diff
@@ -94,3 +94,22 @@ internal struct ComponentEndMarker
 {
     public string? PrerenderId { get; set; }
 }
+
+internal struct ComponentMarkerKey
+{
+    public ComponentMarkerKey(string locationHash, string? formattedComponentKey)
+        => (LocationHash, FormattedComponentKey) = (locationHash, formattedComponentKey);
+
+    // A hash that distinguishes this component from other components in the render tree.
+    // The output should be deterministic between endpoint invocations so that the client
+    // can match up component instances between renders.
+    // The current implementation uses the hashed component type name and its render tree
+    // sequence number.
```

Comment:
It would be slightly more intuitive to me if it used the *parent component's* type name and sequence number, since that uniquely identifies a point in the app's source code. (By parent, I mean the component that rendered the Component rendertree frame).

If this is using the child component's type name and its sequence number in the parent, that could theoretically clash as two different parent types could render the same child type at the same sequence number.

Reply:
That's an interesting idea! For some reason, I had originally thought that there might be cases where the sequence number resets to `0` within a single component, which would also create ambiguities. I'll look into changing this.

Reply:
Another approach would be to use SHA256, which is what @javiercn suggested at one point.

Reply:
@MackinnonBuck sequence numbers are reused within "for" loops I believe, also when using "Region"

Reply:
@javiercn I think that case is okay, because the same component type will be rendered each iteration. The case I'm more worried about is if you had a matching key but mismatching component type. The browser may not know the type of the component it's working with (in Blazor Server at least), so it might accidentally update a component with a mismatching type if the hash is the same.

Reply:
This is a challenge if we use `@key`, because it's easy for devs to use an object id as the key (for example, customer.Id), so we need to make sure that we create something unique enough to make sure that if I render two sets of separate components using `@key` they don't get "merged".

Reply:
Is the concern that using `@key` could result in two components getting merged, where they wouldn't have if a `@key` was not specified? Even if customers use `@key`, we still use the component type/sequence number when verifying that components match up.

Reply:
I'm just trying to explore where it might theoretically go wrong. If you have this component hierarchy:

 * A
   * MyChild (seq=0, key=123, param=val1)

... and then swap to:

 * B
   * MyChild (seq=0, key=123, param=val2)

... then the location hashes for `MyChild` in both cases may be `MyChild:0:123`, so it may decide to retain `MyChild` and update its param from val1 to val2, even though in the developer's mind, these are unrelated instances since they come from completely different parts of their source code.

If the location hashes were `A:0:123` and `B:0:123` then it would not retain.

Reply:
Another solution could be to somehow hash multiple levels of the component heirarchy (up to some maximum depth probably).

Reply:
Can we "lazily" walk up the ancestor tree and combine the info "incrementally". Meaning:
* The first time we encounter a server root component, we calculate a hash up to the parent at each level (that we store)
* Then if a second component gets added, we calculate the hash from the common ancestor to the unique branch of the new component.

What this means is that if we have a hierarchy
```
A -> B -> C -> D
A -> B -> C -> E
```

Where D and E are SSRed components, the first time we find D, we Compute the hash up to C (for that, we compute the hash up to B and combine it with the info from C, storing all intermediate results (In component state))

When we are about to Render E and walk up the hierarchy, we see that we already have a hash for C, and just combine it with E.

Reply:
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.incrementalhash?view=net-7.0 Might be of use here

Reply:
@javiercn I like that a lot!

For the sake of minimizing further changes in this PR, do you think we can consider that as a follow-up item for 8.0 GA?

Reply:
Opened https://github.com/dotnet/aspnetcore/issues/51073 to track this improvement